### PR TITLE
fix `avoid_init_to_null` w/ super initializing non-null defaults

### PR DESCRIPTION
**Pull request by pq**
Fixes: #3349 

I don't love reaching into `Impl` classes so if there's a better way, I'd welcome it!

/cc @bwilkerson

### CODE REVIEW

File: lib/src/rules/avoid_init_to_null.dart [hunk 1]
```diff
@@ -88,8 +90,13 @@ class _Visitor extends SimpleAstVisitor<void> {
   @override
   void visitDefaultFormalParameter(DefaultFormalParameter node) {
     var declaredElement = node.declaredElement;
-    if (declaredElement == null) {
-      return;
+    if (declaredElement == null) return;
+    if (declaredElement is DefaultSuperFormalParameterElementImpl) {
```
Comment (asashour) [inline lib/src/rules/avoid_init_to_null.dart:94]: I guess there is no need to access the implementation, we can use `SuperFormalParameterElement` and `FieldFormalParameterElement`

Comment (bwilkerson) [reply lib/src/rules/avoid_init_to_null.dart:94]: I can't easily see which APIs you need from `DefaultSuperFormalParameterElementImpl` and `DefaultFieldFormalParameterElementImpl` that aren't public APIs, but you might not need to reference them anymore.

You might consider combining the `if`s on lines 93 and 94 as
```
if (declaredElement is! DefaultSuperFormalParameterElementImpl) return;
```
just to make the code a little shorter.

File: lib/src/rules/avoid_init_to_null.dart [hunk 2]
```diff
@@ -88,8 +90,13 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+      var superConstructorParameter = declaredElement.superConstructorParameter;
+      if (superConstructorParameter is DefaultFieldFormalParameterElementImpl) {
+        var defaultValue = superConstructorParameter.defaultValueCode;
+        if (defaultValue != null) return;
```
Comment (bwilkerson) [inline lib/src/rules/avoid_init_to_null.dart:98]: It looks like this is too broad. To confirm or deny, consider adding a test with the following code:
```
class A {
  String? a;
  A({this.a = null});
}
class B extends A {
  B({super.a = null}); // LINT
}
```

**Review by bwilkerson — APPROVED** [review]

File: lib/src/rules/avoid_init_to_null.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/avoid_init_to_null.dart:94]: Thanks!

File: lib/src/rules/avoid_init_to_null.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/avoid_init_to_null.dart:98]: I think the test case you maybe meant was:

```dart
class A {
  String? a;
  A({this.a = ''});
}
class B extends A {
  B({super.a = null});
}
```

and indeed, this is not working as written!

Comment (pq) [reply lib/src/rules/avoid_init_to_null.dart:98]: (Deleted a mistyped example.)

**Pull request merged**